lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -3639,25 +3630,89 @@ bool AdjointGen::performSynthesis(FunctionSynthesisItem item) {
+  if (!originalGenSig)
+    return nullptr;
+  GenericSignatureBuilder builder(original->getASTContext());
+  // Add original generic signature.
+  builder.addGenericSignature(originalGenSig);
+  // Add where clause requirements.
+  auto source =
+      GenericSignatureBuilder::FloatingRequirementSource::forAbstract();
+  for (auto &req : attr->getRequirements())
+    builder.addRequirement(req, source, original->getModule().getSwiftModule());
+  auto canGenericSig = std::move(builder)
+                           .computeGenericSignature(
+                               SourceLoc(), /*allowConcreteGenericParams=*/true)
+                           ->getCanonicalSignature();
+  return canGenericSig;
+}
+
 DifferentiationTask::DifferentiationTask(ADContext &context,
                                          SILFunction *original,
                                          SILDifferentiableAttr *&&attr,
                                          DifferentiationInvoker invoker)
     : context(context), original(original), attr(attr), invoker(invoker) {
+  auto &module = context.getModule();
+  auto originalTy = original->getLoweredFunctionType();
+  auto originalLoc = original->getLocation();
   if (attr->hasJVP()) {
-    jvp = lookUpOrLinkFunction(attr->getJVPName(), context.getModule());
-    assert(jvp);
+    // If attribute specifies JVP name, try to look up JVP in current module.
+    // Otherwise, create an external reference.
+    jvp = module.lookUpFunction(attr->getJVPName());
+    if (!jvp) {
+      auto jvpName = attr->getJVPName();
+      auto *jvpGenericSig = getAutoDiffAssociatedFunctionGenericSignature(attr, original);
+      auto *jvpGenericEnv = jvpGenericSig
+          ? jvpGenericSig->createGenericEnvironment()
+          : nullptr;
+      auto jvpType = originalTy->getAutoDiffAssociatedFunctionType(
+          getIndices().parameters, getIndices().source, 1,
+          AutoDiffAssociatedFunctionKind::JVP, module,
+          LookUpConformanceInModule(module.getSwiftModule()), jvpGenericSig);
+      SILOptFunctionBuilder fb(context.getTransform());
+      jvp = fb.createFunction(SILLinkage::PublicExternal, jvpName, jvpType,
+                              jvpGenericEnv, originalLoc, original->isBare(),
+                              IsNotTransparent, original->isSerialized());
+      // NOTE: Setting debug scope is necessary to prevent crash in TFPartition.
+      jvp->setDebugScope(new (module) SILDebugScope(originalLoc, jvp));
+    }
   }
   if (attr->hasVJP()) {
-    vjp = lookUpOrLinkFunction(attr->getVJPName(), context.getModule());
-    assert(vjp);
+    // If attribute specifies VJP name, try to look up VJP in current module.
+    // Otherwise, create an external reference.
+    vjp = module.lookUpFunction(attr->getVJPName());
+    if (!vjp) {
+      auto vjpName = attr->getVJPName();
+      auto *vjpGenericSig = getAutoDiffAssociatedFunctionGenericSignature(attr, original);
+      auto *vjpGenericEnv = vjpGenericSig
```

Comment:
SILFunctions without a body don't need a GenericEnvironment

Reply:
Done in https://github.com/apple/swift/pull/21973/commits/41716eb58db03ad360f3bafec46e16bf000050d1.